Add the ability for the IPCTestingAPI to report when a sync message fails to be received correctly

gavin-apple opened the pull request:
#### e36c14395c0e952c75c21a9347377024e071a1b8
<pre>
Add the ability for the IPCTestingAPI to report when a sync message fails to be received correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=264506">https://bugs.webkit.org/show_bug.cgi?id=264506</a>
<a href="https://rdar.apple.com/118188347">rdar://118188347</a>

Reviewed by Alex Christensen.

Add the ability for the IPCTestingAPI to report when a sync message
fails to be received correctly. The existing behaviour is that both
failed messages and empty replies return an empty IPC object. This change
records the decoder validity on failure into the MessageSyncReply so the
IPCTestingAPI can raise an exception instead.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchSyncMessage):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::hasMessageTestingError const):
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::setMessageTestingError):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/MessageFlags.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsResultFromReplyDecoder):

Canonical link: <a href="https://commits.webkit.org/270705@main">https://commits.webkit.org/270705@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a072f83886181b6f5160e9f230b24ca2c69a977

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27450 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28272 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23964 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6567 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2207 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22526 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28847 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23489 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23880 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27429 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3285 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1493 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4693 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23237 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6293 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3752 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/MessageFlags.h:
```diff
@@ -32,6 +32,7 @@ enum class MessageFlags : uint8_t {
     DispatchMessageWhenWaitingForUnboundedSyncReply = 1 << 1,
     UseFullySynchronousModeForTesting = 1 << 2,
     MaintainOrderingWithAsyncMessages = 1 << 3,
+    MessageTestingError = 1 << 4,
```
I'd call this something more like SyncMessageDeserializationFailure, and surround it with #if ENABLE(IPC_TESTING_API)

achristensen07 commented on Source/WebKit/Shared/IPCTester.cpp:
```diff
@@ -213,6 +213,12 @@ void IPCTester::syncPing(IPC::Connection&, uint32_t value, CompletionHandler<voi
 
+void IPCTester::syncPingEmptyReply(IPC::Connection&, uint32_t value, CompletionHandler<void()>&& completionHandler)
+{
+    UNUSED_PARAM(value);
```
You can also just omit "value" above.

gavin-apple replied:
I need value to trigger the deserialization error.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2344,6 +2344,11 @@ static JSC::JSObject* jsResultFromReplyDecoder(JSC::JSGlobalObject* globalObject
     auto scope = DECLARE_THROW_SCOPE(vm);
 
+    if (decoder.hasMessageTestingError()) {
+        throwException(globalObject, scope, JSC::createTypeError(globalObject, "Failed to successfully process the message"_s));
```
process->deserialize

gavin-apple committed:
Add the ability for the IPCTestingAPI to report when a sync message fails to be received correctly
https://bugs.webkit.org/show_bug.cgi?id=264506
rdar://118188347

Reviewed by Alex Christensen.

Add the ability for the IPCTestingAPI to report when a sync message
fails to be received correctly. The existing behaviour is that both
failed messages and empty replies return an empty IPC object. This change
records the decoder validity on failure into the MessageSyncReply so the
IPCTestingAPI can raise an exception instead.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::dispatchSyncMessage):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::hasMessageTestingError const):
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::setMessageTestingError):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/MessageFlags.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsResultFromReplyDecoder):

Canonical link: https://commits.webkit.org/270705@main
